[MultiRep HEIC] Add support for custom rendering

pxlcoder opened the pull request:
#### ed687398f954fb809436745aa529879d73c579a2
<pre>
[MultiRep HEIC] Add support for custom rendering
<a href="https://bugs.webkit.org/show_bug.cgi?id=269497">https://bugs.webkit.org/show_bug.cgi?id=269497</a>
<a href="https://rdar.apple.com/122975322">rdar://122975322</a>

Reviewed by Said Abou-Hallawa.

Add custom sizing and baseline support for multi-representation HEIC.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::imageSizeForRenderer const):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::drawMultiRepresentationHEIC):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextCocoa.mm:
(WebCore::GraphicsContext::drawMultiRepresentationHEIC):
* Source/WebCore/platform/graphics/cocoa/MultiRepresentationHEICMetrics.h: Added.
(WebCore::MultiRepresentationHEICMetrics::size const):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::metricsForMultiRepresentationHEIC const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::styleDidChange):
(WebCore::RenderImage::baselinePosition const):
(WebCore::RenderImage::hasAnimatedImage const):
(WebCore::RenderImage::isMultiRepresentationHEIC const):
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/rendering/RenderImage.h:

Canonical link: <a href="https://commits.webkit.org/274890@main">https://commits.webkit.org/274890@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a5ece54826b1a80b7c06005f11bbe50b080a5aa4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40243 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19255 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42621 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42788 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/36332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42550 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22182 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16584 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40817 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16214 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34737 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/14009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14050 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35690 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44066 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36494 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36020 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39769 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12343 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/38048 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16677 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16726 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5333 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16321 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
LGTM

shallawa commented on Source/WebCore/loader/cache/CachedImage.cpp:
```diff
@@ -314,6 +318,11 @@ FloatSize CachedImage::imageSizeForRenderer(const RenderElement* renderer, SizeT
+#if ENABLE(MULTI_REPRESENTATION_HEIC)
+    if (renderer && renderer->isMultiRepresentationHEIC())
+        return sizeForMultiRepresentationHEIC(renderer->style().fontCascade().primaryFont());
+#endif
```
I think it is weird that the renderer can know if the image `isMultiRepresentationHEIC()` but the CachedImage can't. At the same time CachedImage can get the `sizeForMultiRepresentationHEIC()`.

pxlcoder replied:
`CachedImage` is just a resource, it knows nothing about layout/rendering.

This also matches the logic below, which uses `imageOrientation()` from the renderer.

I also have the logic here rather than on the renderer / `RenderImageResource`, since `CachedImage::imageSizeForRenderer` is used in many places.

shallawa commented on Source/WebCore/platform/graphics/cocoa/MultiRepresentationHEICRenderingUtilities.h:
```diff
@@ -0,0 +1,48 @@
+
+FloatSize sizeForMultiRepresentationHEIC(const Font&);
+
+float baselineOffsetForMultiRepresentationHEIC(const Font&);
```
1. I think `baselineOffset` is a little confusing. Does it mean the offset from top of the line or the bottom of the line? In the fonts' world they use `ascent` and `descent`. Can we use one of them based on what you mean by `baselineOffset`?
2. What is the difference between the Font `ascent/descent` and this `baselineOffset`?
3. Can't we name it `descentForCustomGlyph`?

shallawa commented on Source/WebCore/platform/graphics/cocoa/MultiRepresentationHEICRenderingUtilities.h:
```diff
@@ -0,0 +1,48 @@
+
+float baselineOffsetForMultiRepresentationHEIC(const Font&);
+
+ImageDrawResult drawMultiRepresentationHEICIntoContext(GraphicsContext&, const Font&, Image&, const FloatRect&);
```
1. I think this should be a member of `GraphicsContext`.
2. Can't we name it `drawCustomGlyph() `? I just think using `MultiRepresentationHEIC` does not fit with passing a `Font`?

pxlcoder replied:
I know we had discussed moving to `GraphicsContext`, but I'm actually going to move this into `FontCascade`, since that has similar uses of `GraphicsContext`.

pxlcoder replied:
Back to `GraphicsContext` after playing around with this.

shallawa commented on Source/WebCore/platform/graphics/cocoa/MultiRepresentationHEICRenderingUtilities.h:
```diff
@@ -0,0 +1,48 @@
+class Image;
+class LayoutUnit;
+
+FloatSize sizeForMultiRepresentationHEIC(const Font&);
```
1. What is this function getting? the size of what? 
2. It does not take an image as an argument. So why do we have `MultiRepresentationHEIC` in the name? Can't be renamed something like `sizeForCustomGlyph`?
3. If it is font property can't be a method of `Font`?

pxlcoder replied:
This has been refactored to use font metrics.

shallawa commented on Source/WebCore/rendering/RenderImage.cpp:
```diff
@@ -710,7 +732,15 @@ ImageDrawResult RenderImage::paintIntoRect(PaintInfo& paintInfo, const FloatRect
+    auto drawResult = ImageDrawResult::DidNothing;
+#if ENABLE(MULTI_REPRESENTATION_HEIC)
+    if (isMultiRepresentationHEIC())
+        drawResult = drawMultiRepresentationHEICIntoContext(paintInfo.context(), style().fontCascade().primaryFont(), *img, rect);
```
Why is this function not taking the `options` as an argument? Does not it care about the `compositeOperation` and `blendMode`?

pxlcoder replied:
We now use the `options`.

shallawa commented on Source/WebCore/loader/cache/CachedImage.cpp:
```diff
@@ -314,6 +315,13 @@ FloatSize CachedImage::imageSizeForRenderer(const RenderElement* renderer, SizeT
+#if ENABLE(MULTI_REPRESENTATION_HEIC)
+    if (renderer && renderer->isMultiRepresentationHEIC()) {
+        auto metrics = renderer->style().fontCascade().primaryFont().metricsForMultiRepresentationHEIC();
+        return { metrics.ideogramWidth().value_or(0) , metrics.height() };
```
I am not sure why the width can't be retrieved but the height can be. But returning zero for the width does not seem right. Maybe `font.avgCharWidth()` is more appropriate.

shallawa replied:
Oh I see you are using `FontMetrics`. But the meaning of `ideogram` of the font is different from the meaning of `MultiRepresentationHEIC`. See my comment below.

shallawa commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -939,4 +939,15 @@ bool Font::hasAnyComplexColorFormatGlyphs(const GlyphBufferGlyph* glyphs, unsign
+FontMetrics Font::metricsForMultiRepresentationHEIC() const
+{
+    return { };
+}
```
I do not think returning `FontMetrics` is the right type for the purpose of `MultiRepresentationHEIC`. `FontMetrics` contains the metrics for the whole font. So it contains things like `m_lineGap`, `m_lineSpacing` which are related to fonts not the glyphs.

Surprisingly I could not find a type for glyph metrics. I would suggest adding one for your purpose like 

``` 
struct MultiRepresentationHEICMetrics {
    float width;
    float ascent;
    float descent;
    FloatSize size() const { return ... };
};
```

shallawa commented on Source/WebCore/rendering/RenderImage.cpp:
```diff
@@ -284,6 +292,18 @@ LayoutUnit RenderImage::computeReplacedLogicalHeight(std::optional<LayoutUnit> e
+    LayoutUnit offset;
+#if ENABLE(MULTI_REPRESENTATION_HEIC)
+    if (isMultiRepresentationHEIC()) {
+        auto descent = style().fontCascade().primaryFont().metricsForMultiRepresentationHEIC().descent().value_or(0);
```
I think this is very long statement. It has six calls. Can we break it down:

```
auto metrics = style().fontCascade().primaryFont().metricsForMultiRepresentationHEIC();
// then use metrics.descent
```

alanbaradlay commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -2074,6 +2074,17 @@ ImageOrientation RenderElement::imageOrientation() const
+        return false;
+
+    return imageElement->isMultiRepresentationHEIC();
+}
```
any particular reason why this is not on RenderImage (e.g. this is going to evolve to a more generic approach)?

pxlcoder replied:
Not going to evolve to a more generic approach.

I have it here because it's the same way `imageOrientation` is done above, and we have a `RenderElement *` at the call site. Was just hoping to avoid introducing new checks + a virtual method.

alanbaradlay replied:
there's only one call site outside of RenderImage afaict. I don't think it's a good approach to move all the things to the base class and downcast there to specific types (and I don't see why moving it to RenderImage would require to introduce a virtual method instead of just a `dynamicDowncast<RenderImage>`)

pxlcoder replied:
Good point – will make that change.

alanbaradlay replied:
thank you!

pxlcoder committed:
[MultiRep HEIC] Add support for custom rendering
https://bugs.webkit.org/show_bug.cgi?id=269497
rdar://122975322

Reviewed by Said Abou-Hallawa.

Add custom sizing and baseline support for multi-representation HEIC.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/cache/CachedImage.cpp:
(WebCore::CachedImage::imageSizeForRenderer const):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::drawMultiRepresentationHEIC):
* Source/WebCore/platform/graphics/cocoa/GraphicsContextCocoa.mm:
(WebCore::GraphicsContext::drawMultiRepresentationHEIC):
* Source/WebCore/platform/graphics/cocoa/MultiRepresentationHEICMetrics.h: Added.
(WebCore::MultiRepresentationHEICMetrics::size const):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::metricsForMultiRepresentationHEIC const):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::styleDidChange):
(WebCore::RenderImage::baselinePosition const):
(WebCore::RenderImage::hasAnimatedImage const):
(WebCore::RenderImage::isMultiRepresentationHEIC const):
(WebCore::RenderImage::paintIntoRect):
* Source/WebCore/rendering/RenderImage.h:

Canonical link: https://commits.webkit.org/274890@main
